src/database/mod.rs

```diff
@@ -212,12 +214,39 @@ pub(crate) trait DatabaseUtils: Database {
 
 impl<T: Database> DatabaseUtils for T {}
 
+/// A factory trait that builds databases which share underlying configurations and/or storage
+/// paths.
+pub trait DatabaseFactory: Sized {
+    /// Inner type to build
+    type Inner: BatchDatabase;
+
+    /// Builds the defined [`DatabaseFactory::Inner`] type.
+    fn build(
+        &self,
+        descriptor: ExtendedDescriptor,
```

Comment:
I would change this to just take a reference, to avoid cloning the descriptor needlessly